File: app/src/main/java/com/ardabank/aradapay/domain/algorithm/DebtSimplifierEngine.kt
package com.ardabank.aradapay.domain.algorithm

import com.ardabank.aradapay.domain.model.Currency
import java.util.PriorityQueue

data class SimplifiedTransaction(
    val debtorId: String,   // Who needs to pay
    val creditorId: String, // Who receives payment
    val amount: Double,
    val currency: Currency = Currency.TRY
)

object DebtSimplifierEngine {

    /**
     * Minimizes total transaction count using Greedy Flow Algorithm.
     * Takes net balance map (userId -> netBalance) and returns simplified list of transactions.
     */
    fun simplifyDebts(
        netBalances: Map<String, Double>,
        currency: Currency = Currency.TRY
    ): List<SimplifiedTransaction> {
        // Debtor PriorityQueue (largest negative balance first => max debt)
        val debtors = PriorityQueue<Pair<String, Double>>(compareBy { it.second })
        // Creditor PriorityQueue (largest positive balance first => max credit)
        val creditors = PriorityQueue<Pair<String, Double>>(compareByDescending { it.second })

        netBalances.forEach { (userId, balance) ->
            if (balance < -0.01) {
                debtors.add(Pair(userId, -balance)) // Store positive debt amount
            } else if (balance > 0.01) {
                creditors.add(Pair(userId, balance))
            }
        }

        val result = mutableListOf<SimplifiedTransaction>()

        while (debtors.isNotEmpty() && creditors.isNotEmpty()) {
            val debtor = debtors.poll()!!
            val creditor = creditors.poll()!!

            val minAmount = Math.min(debtor.second, creditor.second)
            result.add(
                SimplifiedTransaction(
                    debtorId = debtor.first,
                    creditorId = creditor.first,
                    amount = minAmount,
                    currency = currency
                )
            )

            val remainingDebtorBalance = debtor.second - minAmount
            val remainingCreditorBalance = creditor.second - minAmount

            if (remainingDebtorBalance > 0.01) {
                debtors.add(Pair(debtor.first, remainingDebtorBalance))
            }
            if (remainingCreditorBalance > 0.01) {
                creditors.add(Pair(creditor.first, remainingCreditorBalance))
            }
        }

        return result
    }
}

File: web/src/algorithms/DebtSimplifierEngine.ts
import { Currency } from '../types';

export interface SimplifiedTransaction {
  debtorId: string;   // Who needs to pay
  creditorId: string; // Who receives payment
  amount: number;
  currency: Currency;
}

/**
 * Minimizes total transaction count using Greedy Flow Algorithm.
 * 1:1 Direct Port from Android Kotlin DebtSimplifierEngine.kt
 */
export const DebtSimplifierEngine = {
  simplifyDebts(
    netBalances: Map<string, number>,
    currency: Currency = 'TRY'
  ): SimplifiedTransaction[] {
    // Debtor list (balance < -0.01) => sorted largest debt first
    const debtors: { userId: string; debt: number }[] = [];
    // Creditor list (balance > 0.01) => sorted largest credit first
    const creditors: { userId: string; credit: number }[] = [];

    netBalances.forEach((balance, userId) => {
      if (balance < -0.01) {
        debtors.push({ userId, debt: -balance });
      } else if (balance > 0.01) {
        creditors.push({ userId, credit: balance });
      }
    });

    debtors.sort((a, b) => b.debt - a.debt);
    creditors.sort((a, b) => b.credit - a.credit);

    const result: SimplifiedTransaction[] = [];

    let dIdx = 0;
    let cIdx = 0;

    while (dIdx < debtors.length && cIdx < creditors.length) {
      const debtor = debtors[dIdx];
      const creditor = creditors[cIdx];

      const minAmount = Math.min(debtor.debt, creditor.credit);

      result.push({
        debtorId: debtor.userId,
        creditorId: creditor.userId,
        amount: minAmount,
        currency
      });

      debtor.debt -= minAmount;
      creditor.credit -= minAmount;

      if (debtor.debt < 0.01) {
        dIdx++;
      }
      if (creditor.credit < 0.01) {
        cIdx++;
      }
    }

    return result;
  }
};
